### DISCUSSION TITLE
Conditional variable export in chezmoi.<format>.tmpl throwing undefined error on `chezmoi init`

### DISCUSSION THREAD

**Discussion by robellegate** [twpayne/chezmoi#2907] [Q&A] [answered]
I've been making the switch to chezmoi for my dotfiles, starting with my work configs, but I'm running into the same problem as #2414 where certain variables conditionally set are not available in later parts of the template.

Here's a slightly redacted snippet of my `.chezmoi.yaml.tmpl` file:

```yaml
{{/* The following variables are only available within this template file */}}

{{/* boolean feature tags */}}
{{- $docker := false -}}{{/* true if this machine should have Docker installed */}}
{{- $ephemeral := false -}}{{/* true if this machine is ephemeral, e.g. a cloud or VM instance */}}
{{- $headless := false -}}{{/* true if this machine does not have a screen and keyboard */}}
{{- $remote := false -}}{{/* true if this machine is connected to via SSH */}}
{{- $work := true -}}{{/* true if this machine is used for work, changed from default of false for testing */}}
{{- "" -}}

# ... jumping down to a later part of the template

{{/* variables to set if the machine is used for work */}}
{{- if $work -}}
{{-   $artifactory_host := "artifactory.some-domain.name" -}}
{{-   $artifactory_base := printf "https://%s:443/artifactory" $artifactory_host -}}
{{-   $docker_base := "docker.some-domain.name:443" -}}
{{- end -}}

sourceDir: {{ .chezmoi.sourceDir | quote }}

data:
  # Here we "export" the variables, so we can access them outside this file
  name: {{ $name | quote }}
  email: {{ $email | quote }}

  {{- if $work -}}
  artifactory_host: {{ $artifactory_host | quote }}
  artifactory_base: {{ $artifactory_base | quote }}
  docker_base: {{ $docker_base | quote }}
  {{- end -}}

  docker: {{ $docker }}
  ephemeral: {{ $ephemeral }}
  headless: {{ $headless }}
  remote: {{ $remote }}
  work: {{ $work }}
```

When I run `chezmoi init --source ~/src/git.some-domain.name/robellegate/dotfiles/home`, I encounter the following error message:

```console
chezmoi: template: chezmoi.yaml:61: undefined variable "$artifactory_host"
```

If I remove only the first `{{- if $work -}}` conditional, the `chezmoi init` command successfully evaluates the `promptString` init template functions not pictured but then throws this error:

```console
chezmoi: /Users/robellegate/src/git.some-domain.name/robellegate/dotfiles/home/.chezmoi.yaml.tmpl: yaml: line 21: mapping values are not allowed in this context
```

The template works only after removing both `{{- if $work -}}` conditionals, but obviously this isn't desired when I'm trying to support using the same config for personal and work.

Maybe I can get the outcome I'm looking for by using `.chezmoidata.<format>` or a combination of `.chezmoiignore` and `.chezmoitemplates`, but I'm trying to keep it simple.

I'm using [`.chezmoi.toml.tmpl` from twpayne/dotfiles](https://github.com/twpayne/dotfiles/blob/798742d79da7186267083e3c6b37d2250248ba1d/home/.chezmoi.toml.tmpl) and [`.chezmoi.yaml.tmpl` from gametaro/dotfiles](https://github.com/gametaro/dotfiles/blob/a4a25e514b3392cec706ad075e2a889952ecabb6/home/.chezmoi.yaml.tmpl) as reference in case there's some discrepancy between the TOML and YAML formats for `.chezmoi.<format>.tmpl`.

**Comment by nandalopes** [comment, marked as answer]
You must initialize these variables outside `if` statement, using `:=`. Then inside `if` you update the values with `=`.
